File: 1er_parcial_recuperatorio/java/PrimerParcialRecuperatorio/src/modelos/GestorEnvios.java
package modelos;

import java.util.List;
import java.util.Comparator;

public class GestorEnvios {

    public void gestionarEnvio(List<Transporte> transportes, Paquete paquete, double distancia) {
        System.out.println("Gestionando el envio del paquete pesado:");

        if (transportes.isEmpty()) {
            resolverTransporteNoDisponible();
            return;
        }

        // Filter the available transport options based on weight limits
        List<Transporte> transportesDisponibles = transportes.stream()
            .filter(transporte -> transporte.getPesoMax() >= Math.max(paquete.getPeso(), paquete.calcularPesoVolumetrico()))
            .toList();

        if (transportesDisponibles.isEmpty()) {
            System.out.println("No hay transportes disponibles para este tipo de paquete.");
            return;
        }

        // Select the best transport option based on the given factors
        Transporte mejorTransporte = transportesDisponibles.stream()
            .max(Comparator.comparing(transporte -> transporte.transportarPaquete(distancia, Math.max(paquete.getPeso(), paquete.calcularPesoVolumetrico()))))
            .orElseThrow();

        double factor = mejorTransporte.transportarPaquete(distancia, Math.max(paquete.getPeso(), paquete.calcularPesoVolumetrico()));
        double tiempo = mejorTransporte.tiempoEstimado(distancia);
        double costo = mejorTransporte.costoEstimado(distancia);

        System.out.println("El mejor transporte para el paquete es " + mejorTransporte.getNombre() + " con un factor de " + factor +
            ". Tiempo: " + tiempo + " horas, Costo: " + costo);
    }

    private void resolverTransporteNoDisponible() {
        // Implement logic to handle the case where no transport is available
        System.out.println("No hay transportes disponibles.");
    }
}

File: 1er_parcial_recuperatorio/python/modelos/GestorEnvios.py
class GestorEnvios:
    pass

    def gestionar_envio(self, transportes, paquete, distancia):
        print("Gestionando el envio del paquete pesado:")

        if (len(transportes) == 0):
            self.__resolver_transporte_no_disponible()

        for tranporte_disponible in transportes:
            factor = tranporte_disponible.transportar_paquete(
                paquete, distancia)
            if (factor > 0):
                tiempo = tranporte_disponible.tiempo_estimado(distancia)
                costo = tranporte_disponible.costo_estimado(distancia)
                print(
                    f"El mejor transporte para el paquete es Furgoneta con un "
                    f"factor de {factor}.\n"
                    f"Tiempo: {tiempo} horas, Costo: {costo}"
                )
                break
            else:
                print(
                    f"Paquete demasiado pesado para {tranporte_disponible.nombre}")

        self.__resolver_transporte_no_disponible()

    def __resolver_transporte_no_disponible(self):
        print("No hay transportes disponibles para este tipo de paquete.")
